components/spinner.py

```diff
@@ -3,29 +3,131 @@
 import rev.color
 
 
+class Colour:  # class for storing rgb colours
+    def __init__(self, red: float, green: float, blue: float):
+        self.red = red
+        self.green = green
+        self.blue = blue
+
+    def dist(self, c: Colour) -> int:
+        return (
+            abs(self.red - c.red) + abs(self.green - c.green) + abs(self.blue - c.blue)
+        )
+
+
 class Spinner:
     spinner_motor: wpilib.Spark
-    spinner_solenoid: wpilib.DoubleSolenoid
+    spinner_solenoid: wpilib.Solenoid
     colour_sensor: rev.color.ColorSensorV3
 
-    def __init__(self):
-        pass
+    MAX_COLOUR_DIST = 0.2  # Chosen without experimentation, should be tuned
+    COLOUR_OFFSET = 1  # The offset of the colour sensor, in 45 degree increments
+
+    WHEEL_COLOURS = {  # Colours read in real world
+        "R": Colour(82 / 255, 116 / 255, 55 / 255),
+        "B": Colour(55 / 255, 120 / 255, 80 / 255),
+        "G": Colour(61 / 255, 126 / 255, 66 / 255),
+        "Y": Colour(72 / 255, 131 / 255, 50 / 255),
+    }
 
-    def setup(self):
-        pass
+    WHEEL_ORDER = ["R", "Y", "B", "G"]
 
-    def on_enable(self):
+    WHEEL_SPIN_FACTOR = 0.3
+
+    def __init__(self) -> None:
+        self.rotations = 0
+        self.state = None
+        self.lastCol = None
+        self.required_colour = None
+
+    def on_enable(self) -> None:
         self.spinner_motor.stopMotor()
-        self.spinner_solenoid.set(wpilib.DoubleSolenoid.Value.kReverse)
+        self.spinner_solenoid.set(True)
+        self.rotation = 0
+        self.state = None
+        self.lastCol = None
+        self.required_colour = None
+
+    # Methods to be called by controller
+    def is_complete(self) -> bool:
+        return self.state == None
+
+    def do_rotation_control(self) -> None:  # to start the rotation routine
+        self.state = "rotation"
+        self.rotations = 0
+        self.lastCol = None
+        self.piston_down()
 
-    def execute(self):
-        pass
+    def go_to_colour(self, colour: str) -> None:  # to start the position routine
+        self.state = "position"
+        self.required_colour = colour
+        self.piston_down()
 
-    def read_colour(self):
+    def piston_up(self) -> None:  # as functions to make swapping easier
+        self.spinner_solenoid.set(True)
+
+    def piston_down(self) -> None:
+        self.spinner_solenoid.set(False)
+
+    def read_colour(self) -> tuple:
         return (self.colour_sensor.getColor(), self.colour_sensor.getIR())
 
-    def piston_up(self):
-        self.spinner_solenoid.set(wpilib.DoubleSolenoid.Value.kForward)
+    def get_wheel_dist(
+        self,
+    ) -> int:  # finds the colour wheels distance from required colour in segments
```

Comment:
Why does this simple method signature span multiple lines?

Reply:
ask black

Reply:
Are you going to fix it?